### ISSUE TITLE
training yolo11n-obb for object detection with a class that occupies more than 90% of the image width.

### ISSUE THREAD

**Issue by saeed-ascension** [ultralytics/ultralytics#21471]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Summary
My OBB model predicts correctly during training and validation (model.val()), but during model.predict() inference, output bounding boxes shrink horizontally when the object spans nearly the full image width.

Details
I'm training an OBB model (YOLOv11) where the main object of interest is a long rectangular item that can span ~95% of the image width (on a black background). The object has clear edges and good contrast.

During training and validation, the predicted OBBs closely match ground truth, including full width coverage.

But during inference (model.predict() or model(img)), the predicted OBBs shrink inward from both left and right sides by ~10–15% of the image width. This happens intermittently across images.

This hurts IoU and visual alignment, even though the box is in the right location.

What I’ve tried

rect=True

Increasing box count (anchor matches)

Custom augmentations

Training with rotated and cropped variants

Rechecking padding settings

Manual inference using LetterBox(auto=True, stride=32) to match validation transforms

✔️ If I manually pad the image to 640×640 before inference, the predictions match validation exactly.
❌ But this breaks my overlay pipeline, since I’d then have to undo that padding manually on all results.

Question
Is there a recommended way to prevent this horizontal shrinkage on full-width objects during inference without needing to pad to square or manually reverse transforms?

Would adding a configurable padding ratio (like extra_pad=0.05) inside LetterBox be a clean workaround for these edge-bound object cases?

Let me know if this is expected behavior due to internal stride handling or if it's a mismatch in how val vs predict applies padding.

Thanks!

### Environment

Ultralytics 8.3.151  Python-3.12.10 torch-2.5.1+cu121 CUDA:0 (NVIDIA RTX A500 Laptop GPU, 4096MiB)
Setup complete  (16 CPUs, 30.7 GB RAM, 605.9/952.6 GB disk)

OS                  Windows-11-10.0.26100-SP0
Environment         Windows
Python              3.12.10
Install             pip
Path                ***Ultralytics_Hello_World\yolo_gpu_py312\Lib\site-packages\ultralytics
RAM                 30.65 GB
Disk                605.9/952.6 GB
CPU                 AMD Ryzen 7 PRO 7840HS w/ Radeon 780M Graphics
CPU count           16
GPU                 NVIDIA RTX A500 Laptop GPU, 4096MiB
GPU count           1
CUDA                12.1

numpy                2.1.2>=1.23.0
matplotlib           3.10.3>=3.3.0
opencv-python        4.11.0.86>=4.6.0
pillow               11.0.0>=7.1.2
pyyaml               6.0.2>=5.3.1
requests             2.32.3>=2.23.0
scipy                1.15.3>=1.4.1
torch                2.5.1+cu121>=1.8.0
torch                2.5.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision          0.20.1+cu121>=0.9.0
tqdm                 4.67.1>=4.64.0
psutil               7.0.0
py-cpuinfo           9.0.0
pandas               2.3.0>=1.1.4
ultralytics-thop     2.0.14>=2.0.0

### Minimal Reproducible Example

*

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @saeed-ascension, thank you for your detailed report and for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for additional insights, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as troubleshooting tips.

As this appears to be a 🐛 Bug Report, could you please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)? This will help us reproduce and debug the issue more efficiently.

If you have additional details such as sample images, ground truth annotations, or inference scripts that demonstrate the behavior, including those would be very helpful. Also, if you can clarify your exact Ultralytics YOLO version and share relevant code snippets for your training and inference pipeline, that would further assist our investigation.

Join the Ultralytics community in your preferred space! For real-time support, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For more in-depth discussion, check out [Discourse](https://community.ultralytics.com/), and you can also share with the community on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Please make sure you have upgraded to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to confirm the issue persists on the latest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will also review your issue and assist you soon. Thank you for your patience and for helping us improve YOLO11! 🛠️

**Comment by Y-T-G** [comment]
Did you try with `rect=False`? That adds padding to the inference.

**Comment by saeed-ascension** [comment]
Yes, i've tried rect=False too.

**Comment by Y-T-G** [comment]
I guess you can try reducing `imgsz` during prediction. The issue is that your objects are too big. Lowering `imgsz` will reduce the size of the objects so that it's more easier to cover it completely.

**Comment by saeed-ascension** [comment]
what you are suggesting is to use imgsz = 640 for training and imgsz = 480 for prediction?

**Comment by Y-T-G** [comment]
If you can train at that imgsz, it would be better. But you can also use the existing model to predict at a lower imgsz

**Comment by saeed-ascension** [comment]
I tried this (imgsz = 640 for training and imgsz = 480 for prediction) and I still have the same issue of shrunk boxes.

**Comment by saeed-ascension** [comment]
I am not sure if rect=False is effective, when I manually pad the image to square dimension and pass it to predict, results are ok, when I leave it to model.predict with  rect=False, I ran into this issue.

**Comment by Y-T-G** [comment]
Can you post the code you are using for prediction?

**Comment by saeed-ascension** [comment]
model.predict(
        source  = r"mini_dataset_july20/val/images",
        imgsz   = 640,
        conf    = 0.5,
        save    = True,
        project = r"run",
        name    = "20250720-minidataset",
        exist_ok= True,
        workers = 0,
        device  = 0,
    )

**Comment by saeed-ascension** [comment]
I wonder if I can use model.val() to get OBBs?

**Comment by Y-T-G** [comment]
Can you try imgsz=320?

**Comment by saeed-ascension** [comment]
I fixed this issue by padding my image to a square size (top and bottom, vertical to the longest axis of the object) and then passing it to model.predict. I thought mode.predict is doing this internally, but I think it only stretches the image and doesn't have padding built-in?

**Comment by Y-T-G** [comment]
It pads if you use `rect=False`

**Comment by saeed-ascension** [comment]
I wish it would, it could've saved 12 hours of my time trialing different training rounds. I am sure it doesn't do it properly for me.

**Comment by Y-T-G** [comment]
You can verify it pretty easily. Without `rect=False`, the logs will show that inference wasn't square, i.e. not padded. But with `rect=False`, it's padded to a square input.

<img width="1065" height="524" alt="Image" src="https://github.com/user-attachments/assets/249bca9b-d769-4ff2-99f8-d81f408aeb37" />

**Comment by saeed-ascension** [comment]
It prints the size as square, but results are different than when I pad them separately. @Y-T-G  We can have a chat about it if you would like to?

**Comment by Y-T-G** [comment]
Maybe your padding is different. The Ultralytics padding is performed with LetterBox. `rect=False` just makes `auto=False` here, which pads it to square:

https://github.com/ultralytics/ultralytics/blob/18fee3e7778af501bc69a34c21d768add20e2e43/ultralytics/data/augment.py#L1591

You can even view the padded image if you want:

```python
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
results = model(rect=False)
import cv2
img = cv2.imread("ultralytics/assets/bus.jpg")
cv2.imwrite("padded.jpg",model.predictor.pre_transform([img])[0])
```

I doubt the issue is padding. How exactly are visualizing the predictions? Are you using `result.plot()`?

**Comment by saeed-ascension** [comment]
I tested my images with this script and I get different BBs for padded and original images :
`
        # Run prediction (rect=False ensures square padding)
        results = model.predict(source=img, imgsz=imgsz, rect=False, save=False, verbose=False)

        #  Draw results on **original image**
        plotted_original = results[0].plot(img=img.copy())
        cv2.imwrite(os.path.join(output_original_dir, filename), plotted_original)

        #  Padded input
        padded_img = model.predictor.pre_transform([img])[0]
        results_padded = model.predict(source=padded_img, imgsz=imgsz, rect=False, save=False, verbose=False)

        plotted_padded = results_padded[0].plot(img=padded_img.copy())
        cv2.imwrite(os.path.join(output_padded_dir, filename), plotted_padded)

        print(f"Saved results for {filename} (original + padded)")
`

**Comment by Y-T-G** [comment]
What's different between them?

I tested and the boxes are the same, including the confidences:

**Original**
![original.jpg](https://github.com/user-attachments/assets/b994321e-59f0-4fd6-bb23-3aa1b8420b59)

**Padded**
<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/5a6751c3-9a28-4797-9afe-ed85f38d9447" />

**Comment by saeed-ascension** [comment]
A reminder that this issue only happens for large object that occupies 90% or more of the image width. Unfortunately I can not share the exact image I use, but I will try to replicate this issue with a dummy image.

**Comment by glenn-jocher** [comment]
Thanks for the detailed investigation and debugging! This does sound like a legitimate issue specific to OBB models with very wide objects (90%+ width). The fact that manual padding works but `rect=False` doesn't produce identical results suggests there may be subtle preprocessing differences between validation and prediction modes for OBB.

Since you mentioned you'll try to replicate this with a dummy image, that would be extremely helpful for identifying the root cause. A minimal reproducible example showing the bbox shrinkage with a synthetic wide object would make it much easier to debug and fix.

Given your willingness to submit a PR and the thorough analysis you've done, this looks like a valuable contribution to improve OBB inference consistency. Please share the reproducible example when you have it ready!

**Comment by cqjnovo** [comment]
> A reminder that this issue only happens for large object that occupies 90% or more of the image width. Unfortunately I can not share the exact image I use, but I will try to replicate this issue with a dummy image.提醒一下，这个问题只发生在占据图像宽度 90%或更多的物体上。不幸的是，我无法分享我使用的确切图像，但我将尝试用一张虚拟图像来复现这个问题。

Hello, I have encountered the same issue as you. For large objects whose width exceeds 90% of the image, the detection bounding boxes shrink inward by about 10–15%. However, if I resize the image to 640×640 with padding beforehand and then perform inference, the detection results are correct. Have you already identified the root cause of this problem? I would greatly appreciate it if you could share your findings.

**Comment by Y-T-G** [comment]
@cqjnovo Can you test with this?

`pip uninstall ultralytics -y && pip install git+https://github.com/ultralytics/ultralytics@noclip-obb`

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the PR [Disable box clipping for OBB predictions](https://github.com/ultralytics/ultralytics/pull/21950) by @Y-T-G with contributions from @Laughing-q.

Key changes that may resolve this:
- Updated `scale_boxes` in `ultralytics/utils/ops.py` to:
  - Return unclipped boxes when `xywh=True`
  - Continue clipping only for `xyxy` boxes
- Prevents unintended width/height changes when working with `xywh` (beneficial for OBB workflows)

How to test the fix:
- Install the latest code from main:
  ```bash
  pip install -U git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Or wait for the next official PyPI release and then:
  ```bash
  pip install -U ultralytics
  ```

Feedback requested:
- Please try the updated build and share whether it resolves the issue in your setup. Any minimal reproducer or logs would be helpful if problems persist.

Thanks for reporting this and helping improve the project 🙏
